File: internal/gandalfcore/snapshot/snapshot_test.go
package snapshot_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/qyinm/gandalf/internal/gandalfcore/snapshot"
	"github.com/qyinm/gandalf/internal/gandalfcore/types"
)

func makeSandbox(t *testing.T) (projectPath, homeDir, storeDir string) {
	t.Helper()
	root := t.TempDir()
	projectPath = filepath.Join(root, "project")
	homeDir = filepath.Join(root, "home")
	storeDir = filepath.Join(homeDir, ".gandalf")
	if err := os.MkdirAll(projectPath, 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.MkdirAll(homeDir, 0o755); err != nil {
		t.Fatal(err)
	}
	return projectPath, homeDir, storeDir
}

func TestCaptureCurrentStateProducesManifestWithSchemaVersionAndRFC3339CreatedAt(t *testing.T) {
	t.Parallel()
	projectPath, homeDir, storeDir := makeSandbox(t)
	agent := types.AgentCodex
	scope := types.ScopeUser

	state, err := snapshot.CaptureCurrentState(&types.RuntimeOptions{
		ProjectPath: projectPath,
		HomeDir:     homeDir,
		StoreDir:    storeDir,
		Agent:       &agent,
		Scope:       &scope,
	}, "baseline")
	if err != nil {
		t.Fatalf("CaptureCurrentState: %v", err)
	}
	if state.Snapshot.Manifest.SchemaVersion != "0.1" {
		t.Fatalf("schema version = %q", state.Snapshot.Manifest.SchemaVersion)
	}
	if _, err := time.Parse(time.RFC3339, state.Snapshot.Manifest.CreatedAt); err != nil {
		t.Fatalf("createdAt not RFC3339: %q (%v)", state.Snapshot.Manifest.CreatedAt, err)
	}
	if state.Snapshot.Manifest.Security.RedactionPolicy != "metadata-only" {
		t.Fatalf("redaction policy = %q", state.Snapshot.Manifest.Security.RedactionPolicy)
	}
	if len(state.Snapshot.Content) != 0 {
		t.Fatalf("metadata-only snapshot should have empty content index, got %#v", state.Snapshot.Content)
	}
}

func TestCapturesClaudeUserGlobalSettingsWithContentBacking(t *testing.T) {
	t.Parallel()
	projectPath, homeDir, storeDir := makeSandbox(t)
	claudeDir := filepath.Join(homeDir, ".claude")
	if err := os.MkdirAll(claudeDir, 0o755); err != nil {
		t.Fatal(err)
	}
	settings := `{"permissions":{"allow":["Bash(echo hi)"]}}`
	if err := os.WriteFile(filepath.Join(claudeDir, "settings.json"), []byte(settings), 0o644); err != nil {
		t.Fatal(err)
	}

	agent := types.AgentClaudeCode
	scope := types.ScopeUser
	state, err := snapshot.CaptureCurrentState(&types.RuntimeOptions{
		ProjectPath:    projectPath,
		HomeDir:        homeDir,
		StoreDir:       storeDir,
		Agent:          &agent,
		Scope:          &scope,
		CaptureContent: true,
	}, "claude-baseline")
	if err != nil {
		t.Fatalf("CaptureCurrentState: %v", err)
	}

	foundContent := false
	for _, entry := range state.Snapshot.Content {
		if entry.SourcePath == "~/.claude/settings.json" && entry.CaptureStatus == "captured" {
			foundContent = true
			if entry.Content == nil || *entry.Content != settings {
				t.Fatalf("content = %#v", entry.Content)
			}
			if entry.Checksum == "" || entry.StoragePath == "" {
				t.Fatalf("missing checksum/storage path: %#v", entry)
			}
		}
	}
	if !foundContent {
		t.Fatal("expected Claude settings.json content capture")
	}

	var settingsEvidence *types.DiscoveredItem
	for i := range state.Scan.Evidence {
		item := &state.Scan.Evidence[i]
		if item.Agent == types.AgentClaudeCode &&
			item.Kind == types.KindAgentConfig &&
			item.SourcePath == "~/.claude/settings.json" {
			settingsEvidence = item
			break
		}
	}
	if settingsEvidence == nil {
		t.Fatal("settings evidence not found")
	}
	if len(settingsEvidence.Metadata) == 0 {
		t.Fatal("expected metadata")
	}
	var meta map[string]any
	if err := json.Unmarshal(settingsEvidence.Metadata, &meta); err != nil {
		t.Fatal(err)
	}
	if meta["contentCaptureStatus"] != "captured" {
		t.Fatalf("contentCaptureStatus = %#v", meta["contentCaptureStatus"])
	}
}

func TestSkipsClaudeJSONMetadataOnlyFromContentCapture(t *testing.T) {
	t.Parallel()
	projectPath, homeDir, storeDir := makeSandbox(t)
	if err := os.WriteFile(
		filepath.Join(homeDir, ".claude.json"),
		[]byte(`{"mcpServers":{"docs":{"command":"npx"}}}`),
		0o644,
	); err != nil {
		t.Fatal(err)
	}

	agent := types.AgentClaudeCode
	scope := types.ScopeUser
	state, err := snapshot.CaptureCurrentState(&types.RuntimeOptions{
		ProjectPath:    projectPath,
		HomeDir:        homeDir,
		StoreDir:       storeDir,
		Agent:          &agent,
		Scope:          &scope,
		CaptureContent: true,
	}, "claude-json")
	if err != nil {
		t.Fatalf("CaptureCurrentState: %v", err)
	}
	if len(state.Snapshot.Content) != 0 {
		t.Fatalf("~/.claude.json must stay metadata-only, got %#v", state.Snapshot.Content)
	}
}

File: crates/gandalf-core/tests/snapshot_test.rs
use std::fs;

use gandalf_core::{
    capture_current_state, types::{AgentId, EvidenceKind, EvidenceScope, RuntimeOptions},
};
use tempfile::TempDir;

#[test]
fn captures_claude_user_global_settings_with_content_backing() {
    let root = TempDir::new().expect("temp dir");
    let project_path = root.path().join("project");
    let home_dir = root.path().join("home");
    let store_dir = home_dir.join(".gandalf");
    fs::create_dir_all(&project_path).expect("project dir");
    fs::create_dir_all(home_dir.join(".claude")).expect("claude dir");
    fs::write(
        home_dir.join(".claude/settings.json"),
        r#"{"permissions":{"allow":["Bash(echo hi)"]}}"#,
    )
    .expect("settings");

    let state = capture_current_state(
        &RuntimeOptions {
            project_path: project_path.display().to_string(),
            home_dir: home_dir.display().to_string(),
            store_dir: store_dir.display().to_string(),
            agent: Some(AgentId::ClaudeCode),
            scope: Some(EvidenceScope::User),
            capture_content: Some(true),
        },
        "claude-baseline",
    )
    .expect("capture");

    let content = state.snapshot.content.expect("content entries");
    assert!(
        content.iter().any(|entry| {
            entry.source_path == "~/.claude/settings.json" && entry.capture_status == "captured"
        }),
        "expected Claude settings.json content capture"
    );

    let settings = state
        .scan
        .evidence
        .iter()
        .find(|item| {
            item.agent == AgentId::ClaudeCode
                && item.kind == EvidenceKind::AgentConfig
                && item.source_path == "~/.claude/settings.json"
        })
        .expect("settings evidence");
    assert_eq!(
        settings
            .metadata
            .as_ref()
            .and_then(|m| m.get("contentCaptureStatus"))
            .and_then(|v| v.as_str()),
        Some("captured")
    );
}

#[test]
fn skips_claude_json_metadata_only_from_content_capture() {
    let root = TempDir::new().expect("temp dir");
    let project_path = root.path().join("project");
    let home_dir = root.path().join("home");
    let store_dir = home_dir.join(".gandalf");
    fs::create_dir_all(&project_path).expect("project dir");
    fs::create_dir_all(&home_dir).expect("home dir");
    fs::write(
        home_dir.join(".claude.json"),
        r#"{"mcpServers":{"docs":{"command":"npx"}}}"#,
    )
    .expect("claude json");

    let state = capture_current_state(
        &RuntimeOptions {
            project_path: project_path.display().to_string(),
            home_dir: home_dir.display().to_string(),
            store_dir: store_dir.display().to_string(),
            agent: Some(AgentId::ClaudeCode),
            scope: Some(EvidenceScope::User),
            capture_content: Some(true),
        },
        "claude-json",
    )
    .expect("capture");

    assert!(
        state.snapshot.content.is_none()
            || state
                .snapshot
                .content
                .as_ref()
                .is_some_and(|entries| entries.is_empty()),
        "~/.claude.json must stay metadata-only"
    );
}